Check user authorization in realm api requests

jurgenwerk opened the pull request:
This PR adds authentication and authorization to all our realm's endpoints. If the http method is PUT, PATCH, POST, DELETE, then the request is considered a ✍️ write request, otherwise it's a 📖  read request.

There's a new `checkPermission` function that will:

1. Authenticate the user using the provided JWT 
- If realm is publicly readable or writable it will not require a JWT
- Otherwise it will require a JWT. Will respond with 401 if JWT is missing or invalid.
2. Check whether the current user has a permission to read when reading, and write when writing
- It will respond with 403 if there is no permission to perform the action
- Otherwise it will proceed to process the request

### Notes

1. The realm tests have been refactored to support testing in realms with different permissions
2. The realm permissions config is now passed to the constructor of the realm (as opposed to the previous way of lazily reading it from the environment variable when permissions are checked). This has been changed so that it is more encapsulated and thus easier to test (can easily create different realms with different permissions in the same environment)
3. All the occurences of `REALM_USER_PERMISSIONS="{}"` in our realm start scripts have been removed. `.realms.json.test` and `realms.json.development` files have been introduced (permissions config) for better experience when changing permissions during development or for testing. However, in our deployed environments (not `development` or `test`),  `REALM_USER_PERMISSIONS` is a mandatory ENV variable. This is because in deployed environments it's easier to manage an ENV variable compared to file copying approach 

### Future work

Currently, all our deployed realms are public readable and writable. They will need to stay this way until we add support for realms to be be able to authenticate between each other, otherwise indexing cards from other realms (e.g. linked cards) will fail if there is no permission to read.

jurgenwerk committed:
Check user realm permission when calling realm endpoints

jurgenwerk committed:
wip

jurgenwerk committed:
Merge branch 'main' into cs-6438-check-authorization-of-user-on-each-api-request

jurgenwerk committed:
Check permissions depending on http method

jurgenwerk committed:
Make dev realms world readable and writable

jurgenwerk committed:
RealmPermissions -> RealmPermissionChecker

jurgenwerk committed:
Merge branch 'main' into cs-6438-check-authorization-of-user-on-each-api-request

jurgenwerk committed:
Read permissions for development and test from files, get rid of blank permissions in scripts, simplify and refactor

jurgenwerk committed:
Check write permission when deleting

jurgenwerk committed:
Tests (more to come)

jurgenwerk committed:
Updated pnpm lock

jurgenwerk committed:
Merge branch 'main' into cs-6438-check-authorization-of-user-on-each-api-request

jurgenwerk committed:
Updated pnpm lock

jurgenwerk committed:
Attempt to fix jsonwebtoken deps error

jurgenwerk committed:
Remove .only

jurgenwerk committed:
Linter prefers this

jurgenwerk committed:
Refactor realm tests to support permissioned realms

jurgenwerk committed:
No need to pass the secret because the realm already knows it

lukemelia commented on packages/realm-server/main.ts:
```diff
@@ -9,7 +9,8 @@ import { makeFastBootIndexRunner } from './fastboot';
 import { readFileSync } from 'fs-extra';
 import { shimExternals } from './lib/externals';
-import RealmPermissions from './lib/realm-permissions';
+import { RealmPermissions as RealmPermissionsInterface } from '@cardstack/runtime-common/realm';
```
```suggestion
import { type RealmPermissions as RealmPermissionsInterface } from '@cardstack/runtime-common/realm';
```

lukemelia commented on packages/realm-server/main.ts:
```diff
@@ -157,7 +156,51 @@ let realmPermissions = new RealmPermissions();
     );
-    let permissions = realmPermissions.permissionsForRealm(url);
+
+    let userPermissions = {} as {
```
consider extracting a method/util to keep realm-server/main tidy

jurgenwerk replied:
Good idea, I added this

lukemelia commented on packages/runtime-common/realm-permission-checker.ts:
```diff
@@ -0,0 +1,18 @@
+    let userPermissions = [
+      ...(this.realmPermissions['*'] || []),
+      ...(this.realmPermissions[username] || []),
+    ];
```
Elegant

lukemelia commented on packages/runtime-common/realm.ts:
```diff
@@ -723,17 +733,34 @@ export class Realm {
+
+      if (e instanceof PermissionError) {
+        return new Response(`Not allowed: ${e.message}`, { status: 403 });
+      }
```
I think the language typically used for this is Authentication (who are you?) and Authorization (what are you allowed to do?). Maybe those would be better names than Authorization and Permission?

jurgenwerk replied:
Makes sense. I initially used `AuthorizationError` for authentication errors because of the Authorization header but Authenticaton and Authorization do describe this better. I adjusted this

lukemelia reviewed (approved):
Made a few suggestions. Otherwise good from my POV

jurgenwerk committed:
Merge branch 'main' into cs-6438-check-authorization-of-user-on-each-api-request

jurgenwerk committed:
Extract getting realm permissions logic to a function

jurgenwerk committed:
Adjust types

jurgenwerk commented:
I've addressed all the review comments. 

Also, this PR is somewhat sensitive to the deployment due to ENV var requirements and keeping things work for the world readable realms so I test deployed this branch to our staging. There was a problem that the new error from this PR handling caught successfully - the permissions config (parameter store) was faulty (realm keys were names when they should be URLs). I fixed that issue and redeployed to staging and things seem to be working ok. I've also addressed this issue in production param store.
